assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml

```diff
@@ -81,6 +173,78 @@ spec:
         - name: ssl-certs-host
           mountPath: /etc/ssl/certs
           readOnly: true
+      {{- if eq (int .Values.apiserver.replicas) 1 }}
+        - name: data
+          mountPath: /run/kube-apiserver
+      - name: haproxy
+        image: haproxy:2.1.3-alpine
+        volumeMounts:
+        - name: data
+          mountPath: /run/kube-apiserver
+        command:
+        - /bin/sh
+        - -c
+        - |
+          set -xe
+          export ADDRESS=$(cat /run/kube-apiserver/address)
+          # Make sure initContainer generated kube-apiserver address.
+          if [ -z $ADDRESS ]; then
+            echo "ADDRESS not found"
+            exit 1
+          fi
+          echo "Connecting to $ADDRESS:{{ template "port" . }}"
+          # We use TCP readiness probe and HAProxy does not reject connections if no backend is available,
+          # so we wait until kube-apiserver is available here, so readiness of haproxy container represents
+          # readiness of kube-apiserver, as kube-apiserver cannot have readiness probe set, as it listens
+          # on random IP address.
+          until nc -zv $ADDRESS {{ template "port" . }}; do sleep 1; done
+          echo "Connected"
+          # From https://github.com/docker-library/haproxy/blob/master/Dockerfile-debian.template#L70
+          exec haproxy -f /run/kube-apiserver/haproxy.cfg
+        readinessProbe:
+          tcpSocket:
+            port: {{ template "port" . }}
+      initContainers:
+      - name: config-generator
+        image: haproxy:2.1.3-alpine
+        command:
+        - /bin/sh
+        - -c
+        - |
+          set -xe
+          export ADDRESS="127.$(shuf -i 0-255 -n 1).$(shuf -i 0-255 -n 1).$(shuf -i 1-253 -n 1)"
```

Comment:
`shuf` seems to be part of coreutils. Here we are using the alpine image, that (in a quick search) doesn't install that package, but seems to be a dependency of another installed package: https://github.com/docker-library/repo-info/blob/master/repos/haproxy/local/2.1.3-alpine.md

The debian image (way bigger, probably) does [install coreutils](https://github.com/docker-library/repo-info/blob/master/repos/haproxy/local/2.1.3.md#dpkg-source-package-coreutils830-3).

So, using this command here seems a little fragile (might break when upgrading) making the whole control plane startup to fail if a later image doesn't include coreutils (package dependencies change or the package that has this as dependency is not installed anymore).

Am I missing something? Do we want to depend on this although it seems fragile?

Reply:
I think saying it's fragile is exaggeration here. It works right now and I don't think suddenly HAProxy would stop depending on `coreutils`. Also see the progress on upstream issue regarding SO_REUSEPORT. I really hope we won't keep this solution for long.

Reply:
Ohh, very nice, it seems to be making it's way forward upstream! Awesome, and congrats too! :)

Maybe create a tech-debt issue to either remove this (when SO_REUSEPORT is ready, hopefully in k8s 1.19) or make it more reliable? If you agree to create such an issue, an easy alternative can be this (works on `busybox` sh that alpine uses): `echo $((RANDOM%255))` (or +1 too to mimic the last octet here)

Reply:
OK, changed.